keps/sig-storage/20191117-noupdatable-volumes.md

```diff
@@ -0,0 +1,177 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The most popular and the most convinient way of consuming Secrets and
+ConfigMaps by Pods is consuming it as a file. However, any update to
+a Secret or ConfigMap object is quickly (roughly within a minute)
+reflected in updates of the file mounted for *all* Pods consuming them.
+That means, that a bad update (push) of Secret and/or ConfigMap can
+very quickly break the whole appplication.
+
+The recommended approach for upgrading applications are obviously
+rolling upgrades. For Secrets/ConfigMaps this is translating to
+creating a new object and updating PodTemplate with the reference
+to it. However, that doesn't protect users from outages caused by
+accidental bad updates of existing Secrets and/or ConfigMaps.
+
+Moreover, the feature of updating Secrets/ConfigMaps for running
+Pods is also expensive from scalability/performance point of view.
+Every Kubelet has to watch (default) or periodically poll every
+unique Secret and ConfigMap referenced by any of the Pods it is
+running.
+
+## Motivation
+
+In this KEP, we are proposing introducing an ability to specify that
+particular Secrets/ConfigMaps should not be updated at runtime for a
+given Pods.
+Given there are a lot of users not really taking advantage of automatic
+updates of Secrets/ConfigMaps due to consequences described above, this
+will allow then to:
+- protect themselves better for accidental bad updates that could cause
+  outages of their applications
+- achieve better performance of their cluster thanks to significant
+  reduction of load on apiserver
+
+### Goals
+
+- introduce protection mechanism to avoid outages due to accidental
+  updates of existing Secrets/ConfigMaps
+- improve cluster performance by reducing load on Kubernetes control
+  plance (mostly kube-apiserver) consumed by not-utilized feature
+
+### Non-Goals
+
+- change the default behavior of consumption of Secrets/ConfigMaps
+
+## Proposal
+
+We propose to extend `ConfigMapVolumeSource` and `SecretVolumeSource`
+types with an additional field:
+```go
+  DisableAutoUpdates bool
+```
+
+Based on the value of that field Kubelet will or will not:
+- start a watch (or periodic polling) of a given Secret/ConfigMap
+- perform updates of files mounted to a Pod based on updates of
+  the Kubernetes object
+
+TODO: Note that Kubelet may be restarted in the meantime, so we need to
```

Comment:
Yep, this could further complicate kubelet's pod syncing flow, and the semantics in the API since we'd also need to decide *when* the secret/configmap may receive updates. 

I think immutable secret/configmap is cleaner in the API, and would also justify the complexity in the implementation to support/optimize this. WDYT?

Reply:
Yeah, that's a 2nd strong reason to make this a fact about the configmap/secret and not a fact about the mount.

Reply:
What I was initially thinking is that I don't want to make the Secret/Configmap immutable, to enable e.g. changing finalizers, or OwnerReferences or sth like that.
But I think I agree with pointing my own TODO as an argument against.

What I think though is that we need to make the "conceptual spec" of it immutable and allow mutating of anything in ObjectMeta.
I will update the KEP when I find a bit time (though if you disagree with the above, please shout).

Reply:
Yes, you can't make metadata immutable. You can make the data immutable, though--that's what I had in mind.

Reply:
Updated